installer/bootenv.go

```diff
@@ -151,22 +167,43 @@ func (e *UBootEnv) writeEnvImpl(vars BootVars, separator string) error {
 	log.Debugf("Writing %v to the U-Boot environment, using separator: %s",
 		vars, separator)
 
-	// Make environment update atomic by using fw_setenv "-s" option.
-	setEnvCmd := e.Command("fw_setenv", "-s", "-")
-	pipe, err := setEnvCmd.StdinPipe()
-	if err != nil {
-		log.Errorln("Could not set up pipe to fw_setenv command: ", err)
-		return err
+	var setEnvCmd *exec.Cmd
+	var pipe io.WriteCloser
+	var err error
+	found := false
+	for _, executable := range e.setCommand {
+		// Make environment update atomic by using "-s" option.
+		setEnvCmd = e.Command(executable, "-s", "-")
+		pipe, err = setEnvCmd.StdinPipe()
+		if err != nil {
+			log.Errorln("Could not set up pipe to fw_setenv command: ", err)
+			return err
+		}
+		err = setEnvCmd.Start()
+		if errors.Is(err, exec.ErrNotFound) {
+			log.Debugf("Tried command %q, but it was not found", executable)
+			continue
+		} else if err != nil {
+			log.Errorln("Could not execute fw_setenv: ", err)
+			pipe.Close()
+			if os.Geteuid() != 0 {
```

Comment:
No, it's recommended practice to always try first, and check conditions afterwards. This is because you don't always know what goes on in kernel space. Special arrangements may have taken place to ensure that it works even without root rights, such CAP settings or setuid binaries.